### DISCUSSION TITLE
Operator &== for bitwise operation and enum flag checks

### DISCUSSION THREAD

**Discussion by formix** [dotnet/csharplang#7857] [Language Ideas]
Usage example:

```c#
public enum ExampleStates {
  WillDo = 0x01,
  WontDo = 0x02,
  InterestingIdea = 0x80
}

var state = ExampleStates.WillDo | ExampleStates.InterestingIdea;

// Suggested syntax:
if (state &== ExampleStates.InterestingIdea) {
  Console.WriteLine("Lets think about it some more...");
}

// Logically equivalent to:
if (ExampleStates.WillDo == (state & ExampleStates.WillDo)) {
  Console.WriteLine("Sure lets do that!");
}

// Quite the same but less verbose than:
if (state.HasFlag(EscapeStates.WontDo)) {
  Console.WriteLine("No more C# syntax sugar, thanks.");
}
```

The &== would work with all _int_ types as well.

**Comment by siegfriedpammer** [comment]
I think saving 5 characters compared to the HasFlag invocation (depending on how you format the code) is not going to cut it. Also keep in mind that HasFlag is a well-known method for the runtime, so it can actually produce the same fast code as in-line bit fiddling would produce.

Your example produces the same assembly at runtime in both cases:

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUEZgMwAEApjAK5jkCiAHhGAA4A2lAyghApRTkA3kXLkA6lA4cAIgHtyAXnIAGJqtxwxk+TAQLlajdm2FxASX2UAToISwA5hYAmlCEfUAOVUQC+JBQ45ADCIvg64mTkmAAs5ACyABTMrJw8fAJCKJmUAJThkQD0ReQAMvKOUADGEDIAnlQAjjRQAG511AjkCPIgRFAAZuQpLOxcvPyCAHRSMoZKKkk5U+QAZIxj6ZNZs9Jy8nkFouYxuACcSQBE3DR25FwIAORCLooIABb8AIRXeQDc/iIkXIAVOUSC8QS2FGaQmuWyuWORBK5AAiq0BD0PpRyChWLjgDRulwUEI2rZgPIULjPhAYP1CEMRisBNMABKoABiHAgjlh4wyUxQsz0BkOyNOmAu1wAcoowPJ7iEAMR4+r6CBMPE0RwQGxwbH0gDWIr+gMIfitQA==

To be honest, personally I find the HasFlag invocation more readable than the bit fiddling, even though I am a programmer with 22+ years of experience with C and C-like languages.

**Reply by formix** [reply]
Thanks for your feedbak siegfriedpammer. That is a good reason to not do it for sure. But I felt that is was quite similar to the other assignation operator mashup patterns out there (+=, &=, |=, etc) to be it's own thing.

The goal is not to make the code faster or more optimized but to have convenient and close to the existing shorter way to squash an equality test to a bitwise & with a clear (and maybe narrow) use case.

That suggestion is more about syntax sugar. That is appreciated by some people who like to have more compact code without loosing on expressiveness. Not intended for the fist exercise in a programming tutorial for sure.

Now how you define 'expressiveness' and if you consider &== to be expressive may vary based on your personal taste, experience and knowledge.

**Reply by siegfriedpammer** [reply]
I don't want to sound harsh and unfair, but how are you going to defend the merit of this proposal? 

Like I said: How are 5 characters less typing and personal preference/taste going to justify the money and countless person hours spent to realize this feature?

It requires:
* countless hours spent in meetings to make sure this feature is correctly specified and fits into the language.
* time to implement this feature in the compiler, write tests, make sure it works in every aspect of the ecosystem: the compiler, the IDE. There are countless tools involved.

It adds additional complexity to the language. Language features are not opt-in for reading the source code of a project. If one developer uses this feature, all the other devs will be forced to understand it too. 

Have you ever heard of - 100 points? https://learn.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points

**Reply by formix** [reply]
I do not want to defend its merit. It's a proposal and if it makes sense, the team in charge of the evolution of the language will decide by themselve what to do about it. I explained why I thought it might be interesting and how to use it. You shared your opinion about why you don't think it would be worthwhile to go ahead with that. It is all fair and good. From now on, I don't care. The process to decide if the feature will go through will unwrap and whoever in charge may close that ticket with the resolution that fits the action taken.

Still want to thank you for that last comment, I didn't know about that minus 100 points article. I will take some time te read it through.

Peace, out!
